doc/main/scala/lms/core/backend.md

```diff
@@ -128,25 +114,33 @@ x = 10
 print(x)
 ```
 
-In this small example, the node `x = 10` is not *data-depended* by the print statement.
+In this example, the node `x = 10` is not *data-depended* by the print statement.
 However, it is *control-depended* because the value of `x` is mutated by it. However,
 the `x = 3` node can be removed without breaking the code semantics. The order of
-graph construction (the fact that user did `x = 3` before `x = 10`) is the base of
-our control dependency analysis.
+graph construction (the fact that the user did `x = 3` before `x = 10`) is the base of our control dependency analysis.
+
+### Effects and EffectSummary
+
+How can we obtain data and control dependency information from the source program? As we mentioned before, data dependency is already captured by the definition of `Node`. However, it is not clear how to collect control dependencies. It turns out that we can track the *effects* of nodes and blocks and then compute control dependencies from these effects. To make our analysis more fine-grained (so that more interesting code reordering and optimization can be introduced), the first step is to realize that there are different kinds of effects. The effect of printing is clearly not so related to the effect of variable reading. Thus there should be no scheduling enforcement between them.
+In LMS IR, we track *read* and *write* effects. They are categorized in the following cases:
```

Comment:
I guess it would be good to point out that these "phantom" effect keys (such as STORE/CTRL) can be user-defined too.